### PR TITLE
octree updates

### LINKED ISSUE #1178: Store the point indices that belong to a leaf/internal node in Octree
**Is your feature request related to a problem? Please describe.**

Would it be possible to store the point indices that belong to a leaf/internal node in Octree? Thank you very much.

**Describe the solution you'd like**
Could you either store the info as a field of OctreeNodeInfo or as a field of OctreeInternalNode?

**Describe alternatives you've considered**
I have gone through the python and cpp code and didn't see it store the indices any where.

**Additional context**

### LINKED ISSUE #2681: Query for multiple points in Octree
**Is your feature request related to a problem? Please describe.**
Is it possible to access points inside a leaf node directly?

**Describe the solution you'd like**
A method returning points (either indices or coordinates) belonging to a leaf node. 
i.e: given leaf node N-> return np.array([3,1], [3,1], ...)

**Describe alternatives you've considered**
Possible alternative could be quering all points in the respective point cloud thus proving link between point and leafnode, using the method locate_leaf_node(from Octree), but to slow on larger point clouds. 

Another approach is to traverse the tree, when reaching a leaf node get coordinates then filtering point cloud based on these coordinates. But unable to acess coordinates of leafNodes when traversing, as it returns ->OctreeColorLeafNode with color [0, 0, 0]

**A clear and concise description of any alternative solutions or features you've considered.**
Simply using VoxelGrid instead, where I can query on a list of points.

I might just have overlooked the method or I am blind, but can not seem to find a meaningful way to get the list of points belonging to a leaf node.

Thanks.
/j

### PR DESCRIPTION
**Pull request by plusk01**
Changes to `Octree` include:
- new `OctreePointColorLeafNode` (subclass of `OctreeColorLeafNode`) to allow storing of point cloud point indices in each leaf node
- Previously only `OctreeLeafNode` objects had `f_init` and `f_update` functions, which provide flexibility to the user for custom leaf objects. This has been extended to `OctreeInternalNode` objects with `fi_init` and `fi_update` functions (for leaf nodes, renamed to `fl_init`, `fl_update`). For API backward compatibility, if these functions are omitted, the default behavior (i.e., nothing) is used.
-  new `OctreeInternalPointNode` (subclass of `OctreeInternalNode`) to allow internal nodes to store the point cloud point indices of all of its children (useful for searching / processing regions of the point cloud)
- Allow early stopping in `Octree::Traverse` method---For a point cloud with ~50k points and a `max_depth` of 9, this makes `LocateLeafNode` run at ~7 usec. Without early stopping, this search visits every node and takes ~5000 usec.
- C++ and Python examples (via a .ipynb tutorial), added to tutorial docs next to `voxelization`

Fixes #1178, fixes #2681

---
This change is [Reviewable](https://reviewable.io/reviews/intel-isl/open3d/2833)

### CODE REVIEW

**Comment by plusk01** [comment]
I also tried out @griegler's suggestion (https://github.com/intel-isl/Open3D/issues/1178#issuecomment-540416136) to help users understand the flexibility of the `Octree` class, however, I kept facing inheritance / downcasting walls with pybind (e.g., see attached). It seems related to #572. If anyone has any hints, I'd be happy to try it out.

octree_advanced_error

File: examples/python/geometry/octree.ipynb [hunk 1]
```diff
@@ -0,0 +1,314 @@
+{
```
Comment (yxlao) [inline examples/python/geometry/octree.ipynb:1]: Please clear all outputs for cells before checking-in the notebook. The CI will run the notebook automatically.

**Review by yxlao — CHANGES_REQUESTED** [review]

**Comment by yxlao** [comment]
> I also tried out @griegler's suggestion ([#1178 (comment)](https://github.com/intel-isl/Open3D/issues/1178#issuecomment-540416136)) to help users understand the flexibility of the `Octree` class, however, I kept facing inheritance / downcasting walls with pybind (e.g., see attached). It seems related to #572. If anyone has any hints, I'd be happy to try it out.

Regarding this, I don't have a readily available solution. We used python inheritance before, e.g. https://github.com/intel-isl/Open3D/blob/45fc6d3e4e4bb43e945f7297bb8e7788f310b2d2/python/open3d/core.py#L209, but it was for a simpler case without the trampoline class. The inheritance structure and trampoline class make `OctreeColorLeafNode`'s python inheritance more tricky.

File: cpp/open3d/geometry/Octree.cpp [hunk 2]
```diff
@@ -417,48 +578,71 @@ void Octree::ConvertFromPointCloud(const geometry::PointCloud& point_cloud,
...
 void Octree::InsertPoint(
         const Eigen::Vector3d& point,
-        const std::function<std::shared_ptr<OctreeLeafNode>()>& f_init,
-        const std::function<void(std::shared_ptr<OctreeLeafNode>)>& f_update) {
+        const std::function<std::shared_ptr<OctreeLeafNode>()>& fl_init,
+        const std::function<void(std::shared_ptr<OctreeLeafNode>)>& fl_update,
+        const std::function<std::shared_ptr<OctreeInternalNode>()>& fi_init,
+        const std::function<void(std::shared_ptr<OctreeInternalNode>)>&
+                fi_update) {
+    // if missing, create basic internal node init and update functions
```
[the remark this answers is not in the document]
Comment (plusk01) [reply cpp/open3d/geometry/Octree.cpp:595]: capitalization?

File: cpp/open3d/geometry/Octree.cpp [hunk 3]
```diff
@@ -180,6 +266,78 @@ bool OctreeColorLeafNode::ConvertFromJsonValue(const Json::Value& value) {
...
+    cloned_node->indices_ = indices_;
+    return cloned_node;
+}
+
+bool OctreePointColorLeafNode::operator==(const OctreeLeafNode& that) const {
+    try {
+        const OctreePointColorLeafNode& that_point_color_node =
+                dynamic_cast<const OctreePointColorLeafNode&>(that);
+
+        return this->indices_.size() == (that_point_color_node.indices_.size());
```
[the remark this answers is not in the document]
Comment (plusk01) [reply cpp/open3d/geometry/Octree.cpp:307]: Good catch. I've added `color_` comparison and `indices == indices`.

An `unordered_set` would be useful for the indices comparison step, but I chose `vector` for simplicity and to better interface with `PointCloud::SelectByIndex(const std::vector<size_t> &indices, bool invert = false)`. Further, I'm not sure where `LeafNode::operator==` is being used, so the cost of `vector` comparison may be small (maybe I'm wrong?)

File: cpp/open3d/geometry/Octree.cpp [hunk 4]
```diff
@@ -180,6 +266,78 @@ bool OctreeColorLeafNode::ConvertFromJsonValue(const Json::Value& value) {
...
+    return [idx,
+            color](std::shared_ptr<geometry::OctreeLeafNode> node) -> void {
+        if (auto point_color_leaf_node = std::dynamic_pointer_cast<
+                    geometry::OctreePointColorLeafNode>(node)) {
+            OctreeColorLeafNode::GetUpdateFunction(color)(
+                    point_color_leaf_node);
+            point_color_leaf_node->indices_.push_back(idx);
+        } else {
+            utility::LogError(
+                    "Internal error: leaf node must be OctreeLeafNode");
```
[the remark this answers is not in the document]
Comment (plusk01) [reply cpp/open3d/geometry/Octree.cpp:290]: Oops, I meant `OctreePointColorLeafNode`

**Comment by plusk01** [comment]
rebased and force pushed to see if that fixes Ubuntu OpenBLAS CI arm64 build (`pillow` issue)

**Comment by yxlao** [comment]
Thanks @plusk01. 

Right, recent updates from `pillow` broke our arm64 CI. It has been resolved in `master`.

**Pull request merged**